File: 08_leader/dominator.py
def solution(A):
  n = len(A)
  size = 0

  #find candidate
  for i in xrange(n):
    if size == 0:
      val = A[i]
      size = 1
    elif val == A[i]:
      size += 1
    else:
      size -= 1

  #verify candidate
  cnt = 0
  for i in xrange(n):
    if val == A[i]:
      loc = i
      cnt += 1
  if cnt > n // 2:
    return loc
  else:
    return -1

File: 08_leader/dominator.cpp
#include "vector"

using namespace std;

int solution(vector<int> &a) {
  int ret = -1;
  int n = a.size();
  int candidate;
  int size = 0;
  int cnt = 0;
  for (int i = 0; i < n; i++) {
    if (size == 0) {
      size++;
      candidate = a[i];
    }
    else if (candidate == a[i]) {
      size++;
    }
    else {
      size--;
    }
  }
  for (int i = 0; i < n; i++) {
    if (a[i] == candidate) {
      ret = i;
      cnt++;
    }
  }
  if (cnt > n / 2) {
    return ret;
  }
  else {
    return -1;
  }
}
